oximeter/db/src/oxql/ast/table_ops/filter.rs

```diff
@@ -30,33 +30,212 @@ use std::fmt;
 /// expression, such as: `filter hostname == "foo" || (hostname == "bar"
 /// && id == "baz")`.
 #[derive(Clone, Debug, PartialEq)]
-pub enum Filter {
-    /// An individual filter atom, like `foo == "bar"`.
-    Atom(FilterAtom),
-    /// A filtering expression combining multiple filters.
-    Expr(FilterExpr),
+pub struct Filter {
+    /// True if the whole expression is negated.
+    pub negated: bool,
+    /// The contained filtering expression, which may contain many expressions
+    /// joined by logical operators.
+    pub expr: FilterExpr,
+}
+
+impl fmt::Display for Filter {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}({})", if self.negated { "!" } else { "" }, self.expr,)
+    }
+}
+
+impl core::str::FromStr for Filter {
+    type Err = Error;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        crate::oxql::ast::grammar::query_parser::filter_expr(s)
+            .map_err(|e| anyhow::anyhow!("invalid filter expression: {e}"))
+    }
 }
 
 impl Filter {
-    // Merge this filter with another one.
-    pub(crate) fn merge(&self, other: &Filter) -> Self {
-        Filter::Expr(FilterExpr {
-            left: Box::new(self.clone()),
-            op: LogicalOp::And,
-            right: Box::new(other.clone()),
-        })
+    /// Return the negation of this filter.
+    pub fn negate(&self) -> Filter {
+        Self { negated: !self.negated, ..self.clone() }
+    }
+
+    /// Split the filter at top-level disjunctions.
+    ///
+    /// This is likely only useful after simplifying to DNF with
+    /// `simplify_to_dnf()`.
+    pub fn flatten_disjunctions(&self) -> Vec<Self> {
+        let mut out = vec![];
+        self.flatten_disjunctions_inner(&mut out);
+        out
+    }
+
+    fn flatten_disjunctions_inner(&self, dis: &mut Vec<Self>) {
+        // Recursion is only needed if this is an OR expression. In that case,
```

Comment:
Yes, exactly. The tree is always right-deep, and the root node's "position" in the original expression moves depending on how the operator precedence is parsed. It's kind of an unfortunate coupling, in that if we change the parser, this might change too, but also a useful simplification. Maybe another reason to write a manual parser....